Elastic Kibana Version Scheme Fix

philippkahr opened the pull request:
Hallo

Sorry for the confusion. I opened a pull request at johnkeates formula and not here.

first of all, there is a problem regarding the versions.yaml. It is a dict. I cannot install 5.5.1 cause it has only version "5" in their dict. I had to add "5.5.1", I know this is not a clean solution, but I could not figure out, how I could do a "match only first number of repoVersion Variable and check if dict exists".

I also added - version: {{ kibana.repoVersion }} in the install.sls

proof
[INFO ] Executing command ['systemd-run', '--scope', 'apt-get', '-q', '-y', '-o', 'DPkg::Options::=--force-confold', '-o', 'DPkg::Options::=--force-confdef', 'install', 'kibana=5.5.1'] in directory '/root'

Commit:
added support for version 5.5.1

Commit:
added support for versions via repoVersion Variable in install.sls

Commit:
#13

philippkahr commented:
Added fix for #13 in config.sls should not be needed if package was successfully installed via apt-get / yum.

johnkeates commented:
It's currently not passing CI, could you check that your changes don't break and try again? It's currently breaking (as expected) on the version numbers. The problem with the PR is that in it's current form it will break everyone's setup.

philippkahr commented:
@johnkeates I get that. I am sorry, I should have checked it with my local build with different version. Hmm, I am not aware how you install a specific version with your formula, while using the repo. I'll think of something.

Commit:
Added if conditions for version 5

Commit:
Added if conditions for repoVersion

philippkahr committed:
Revert "Added if conditions for repoVersion"

This reverts commit 67772625cd8f8eea2cb10fd38368af9c63c63b6d.

philippkahr committed:
Forget Map Jinja

Travis CI Build not possible due to missing variables

philippkahr committed:
Created an endless loop. Required Map jinja in versions.yaml which was required in map.jinja.

philippkahr committed:
Missed Paranthesis

philippkahr committed:
Paranthesis error

philippkahr committed:
Trying

philippkahr committed:
Fix

johnkeates commented:
@philippkahr no worries, it happens all the time. I'm not sure what the best solution is, but it definitely has to pass first.

I think we should have some sort of version switch between the older (pre 5) releases that people still use and the newer ones. Elastic makes some other products that have some pretty bad breaking changes too, so we may need to implement some sort of major switching scheme in there as well.

Commit:
Fixed the issue with Version 5. Regarding use of minor Versions.

Commit:
Well in versions.yaml the dict has to contain a plain version 5. Hmm

Commit:
.

Commit:
{% endif %} missing

Commit:
Added version fix for "5." cause installing `apt-get install kibana=5` wouldnt found a packet.

philippkahr commented:
@johnkeates I did a bit of tricking with those version numbers, as Elastic makes it really hard to automate that kind of stuff with their different repos.

First of all I added a fix for #13 in config.sls with that https://github.com/saltstack-formulas/kibana-formula/pull/16/commits/bc49101450e352ce37900c2062bc307681341425 commit. 

Second, I added a match for the variable repoVersion if it starts with "5." due to Elastic not allowing to install "kibana=5" with apt-get. My solution is not beautiful. It works the following way.:
in versions.yaml it checks if "pillar.get[repoVersion]" is not null, then it creates a new variable versionRepo, if that versionRepo starts with "5." it will create an entry in versions.yaml so that the dict contains the version for installing. In install.sls if "pillar.get[repoVersion]" startswith "5." it will add the version tag to the apt-get installation.

I am sorry for the noise with all the commits. I am a bit new to testing with travis and kitchen. My local build had some troubles with those gems, so I had to create a lot of commits for github so that travis can tests them.

blbradley commented on kibana/config.sls:
```diff
@@ -3,6 +3,15 @@
 
 {% from "kibana/map.jinja" import kibana with context %}
 
+{% if kibana.configfile == '/etc/kibana/kibana.yml' %}
```
If you did `makedirs: True` on the state below this, you won't need this new block. Also, this formula already uses a different config directory. So, I don't think you should check against a config different in a different path.

philippkahr replied:
My thoughts behind the if clause are that the config file paths are different for version 5 and 4.

> equal or greater 5 /etc/kibana/
> less than 5 /opt/kibana/

therefore I needed to check if it's version 5 and that is done via the configfile path. As someone has to set it in their pillar, else the config file will be deployed to 

> /opt/kibana/config

due to defaults.yaml.

Checking against the version is a bit difficult, since repoVersion and sourceVersion might not be set correctly in the pillar and the defaults.yaml would apply.

blbradley replied:
Then why not `{% if kibana.version.startswith('5') %}`?

blbradley replied:
I also find it odd that the config directory is not created by the package.

blbradley replied:
That's it! `kibana-pkg` should run before `kibana-config`. You can do this with `require` or `require_in`.

blbradley commented on kibana/install.sls:
```diff
@@ -15,4 +15,7 @@ include:
 kibana-pkg:
   pkg.installed:
     - name: {{ kibana.pkg }}
+{% if kibana.repoVersion.startswith('5.') %}
```
You should stick to the style of the rest of the file unless you have a reason. i.e. Your `{%` should be `{%-`.

edit: syntax error

philippkahr replied:
@blbradley Hey there, I really don't understand the `-` operator in jinja2.

Documentation says 

> You can also strip whitespace in templates by hand. If you add a minus sign (-) to the start or end of a block (e.g. a For tag), a comment, or a variable expression, the whitespaces before or after that block will be removed:

To be honest, I am not a programmer, I am somewhat of a beginner in python and in java, so the concept of "whitespace" is relatively new to me. I'll edit all files and add the `-` operator.

blbradley replied:
Hi! 😄 

For SaltStack in Jinja, that operator is commonly used to control if an newline appears or not for a conditional statement.

Does that clear up your knowledge? Feel free to follow up with questions.

philippkahr replied:
@blbradley ah so without `-` operator the final rendered file should look like

```
kibana-pkg:
     pkg.installed:
         - name: kibana

         - version: 5.1.1
```

and with the `-` operator

```
kibana-pkg:
     pkg.installed:
         - name: kibana
         - version: 5.1.1
```

blbradley replied:
Correct!

blbradley commented on kibana/versions.yaml:
```diff
@@ -1,6 +1,16 @@
+  {% set versionRepo = salt['pillar.get']('kibana:repoVersion') %}
+  {% if versionRepo.startswith('5.') %}
+{{ versionRepo }}:
+  repo_url: https://artifacts.elastic.co/packages/5.x/apt
```
I wonder if there is a way the formula can access the data under dict key `5` while still setting your full version number 5.5.1. Clearly, the repo for all versions of ES 5 is the same.

johnkeates replied:
I think the version 5 thing will work out so far, but Elastic will probably mess this up again and then we'll have to cover yet another case. The good thing about the way it's implemented now is that it works for older and current versions alike.

We should probable put in a memo somewhere so that when future versions get released, anyone interested in maintaining knows what's up.

philippkahr replied:
@blbradley The only way to fix this would be to do some magic in the map.jinja where the repo dict is used. I don't believe I am good enough to get this done in a short amount of time, so this PR can go through.

I created a issue regarding that #17

blbradley replied:
I agree! This magic you created will work.

blbradley reviewed (commented):
Thanks for the code! Here's some pointers.

johnkeates commented:
I'll merge in a little while if no other objections arise and you have the small edits in :)

Commit:
Add - operator for install.sls

Commit:
Added a disclaimer for the version.yaml

Commit:
config.sls added require_in

blbradley commented on kibana/config.sls:
```diff
@@ -21,3 +25,5 @@ kibana-config:
     - group: root
     - dataset: {{ kibana.config }}
     - formatter: "YAML"
+    - require_in:
```
This should be `require`, and then you won't need your additional state above this one: `kibana-create-config-folder`.

philippkahr replied:
@blbradley oh yes sorry, you are right. I just did a test install in a docker container with the source package. I thought that if you install it from source it may not create the required folder. But it does create it. I guess #13 is some weird issue.

blbradley replied:
No problem! I appreciate your patience and acceptance of my help. I saw the storm of commits this morning in my inbox (not an issue at all). Felt like I could use this as a teaching experience.

philippkahr replied:
Haha, yeah I am really sorry for the noise that the commits created. I thought I had to create a commit for each change I did, so everyone could see my trial and error experience, especially regarding the use of jinja2 "startswith" stuff. I now get that I'll only create a commit if it's building and working in my local kitchen setup.

I am still struggeling with the information I should put into commits.

Thank you very much for time and also @johnkeates as I did a whole lot of mess here.

blbradley replied:
@philippkahr Don't worry about it too much; it's more of an art than a strict set of rules.

Commit:
@blbradley suggested require instead of require_in.

blbradley committed:
logical error

blbradley commented:
I made a small fix to pass the test suite.

blbradley commented:
@philippkahr Care to change the title to match this? This became a bugfix on using Elastic's version scheme.

philippkahr reviewed (commented):
Why is it needed to add "kibana.install" when I included the install.sls at top? I know that the built has failed, but I don't get it.

blbradley commented:
When you add this formula, the path from the root to the sls file is `kibana/install.sls` which corresponds to `kibana.install`.

@philippkahr You want to `Squash and Merge`? Feel free to add your own message.

blbradley commented:
I find it odd that your commit with `require_in` passed. Testing doesn't cover everything! 🎐

philippkahr commented:
@blbradley maybe `require_in` resolves the `kibana/install.sls` to `install.sls`. That's why I was confused, because `require` didn't work with `install.sls`
